File: integration-test/java/jee/src/main/java/com/qualityminds/lazyval/integration/boundary/rest/UnsupportedPersistenceTypeException.java
package com.qualityminds.lazyval.integration.boundary.rest;

import com.qualityminds.lazyval.integration.boundary.rest.model.PersistenceType;

/**
 * Thrown when a request names a {@link PersistenceType} that is valid per the API contract but
 * not wired in this deployment. Extends {@link UnsupportedOperationException} for readability,
 * but the 501 mapper binds to this subtype only — mapping the JDK type would silently turn
 * genuine bugs (mutating an immutable collection, say) into "not implemented" responses.
 */
public class UnsupportedPersistenceTypeException extends UnsupportedOperationException {

    private final PersistenceType persistenceType;

    public UnsupportedPersistenceTypeException(PersistenceType persistenceType) {
        super("Persistence type " + persistenceType + " is not implemented by this application");
        this.persistenceType = persistenceType;
    }

    public PersistenceType getPersistenceType() {
        return persistenceType;
    }
}

File: integration-test/kotlin/jee/src/main/kotlin/com/qualityminds/lazyval/integration/boundary/rest/UnsupportedPersistenceTypeException.kt
package com.qualityminds.lazyval.integration.boundary.rest

import com.qualityminds.lazyval.integration.boundary.rest.model.PersistenceType

/**
 * Thrown when a request names a [PersistenceType] that is valid per the API contract but not wired
 * in this deployment. Extends [UnsupportedOperationException] for readability, but the 501 mapper
 * binds to this subtype only — mapping the stdlib type would silently turn genuine bugs into
 * "not implemented" responses.
 */
class UnsupportedPersistenceTypeException(val persistenceType: PersistenceType) :
    UnsupportedOperationException("Persistence type $persistenceType is not implemented by this application")
